🎉 Source Notion: Adds pagination to Users stream

lgomezm opened the pull request:
## What
* Notion `Users` stream does not handle pagination. That is, it only outputs the first page of user records, which defaults to 100 (https://developers.notion.com/reference/pagination).

## How
* Adds correct handling of pagination for that kind of stream.
* Overwrites the `request_params` stream method.

## Recommended reading order
1. `streams.py`
2. `test_streams.py`

## 🚨 User Impact 🚨
Are there any breaking changes? What is the end result perceived by the user? If yes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

*Put your unit tests output here.*

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

itaseskii commented on airbyte-integrations/connectors/source-notion/unit_tests/test_streams.py:
```diff
@@ -76,3 +77,61 @@ def test_backoff_time(patch_base_class):
+
+    # When getting pages after the first pull.
+    inputs = {"stream_slice": None, "stream_state": None, "next_page_token": {"next_cursor": "123"}}
+    expected_params = {"start_cursor": "123", "page_size": 100}
```
Instead of using a dummy **_start_cursor_** value maybe use a self documenting value that will be provided from **_Notion_** such as  **_"next_cursor": "fe2cc560-036c-44cd-90e8-294d5a74cebc"_**

lgomezm replied:
You're right. Updated in [fda250e](https://github.com/airbytehq/airbyte/pull/11452/commits/fda250ecf078624eff22eb745c3eaa7337c947ce).

itaseskii commented on airbyte-integrations/connectors/source-notion/unit_tests/test_streams.py:
```diff
@@ -76,3 +77,61 @@ def test_backoff_time(patch_base_class):
+    """
+
+    response_body = {
+        "object": "list",
```
In the original response from **_Notion_** both **_object_** and **_type_** are part of the **_results_** array. In my opinion It would be better if the tests are kept consistent with the actual implementation
![image](https://user-images.githubusercontent.com/10071532/160299110-2606bb0c-ba3d-4f68-84cc-9a2c737d2858.png)

lgomezm replied:
Done in [fda250e](https://github.com/airbytehq/airbyte/pull/11452/commits/fda250ecf078624eff22eb745c3eaa7337c947ce).

itaseskii commented on airbyte-integrations/connectors/source-notion/unit_tests/test_streams.py:
```diff
@@ -76,3 +77,61 @@ def test_backoff_time(patch_base_class):
+        "has_more": True,
+        "type": "user"
+    }
+    url = "https://api.notion.com/v1/users?page_size=100"
```
Is there a need to declare the **_url_** variable? couldn't the url be passed directly to **_requests_mock.get()_**?

lgomezm replied:
Moved to `requests_mock.get()` in [fda250e](https://github.com/airbytehq/airbyte/pull/11452/commits/fda250ecf078624eff22eb745c3eaa7337c947ce).

itaseskii commented on airbyte-integrations/connectors/source-notion/unit_tests/test_streams.py:
```diff
@@ -76,3 +77,61 @@ def test_backoff_time(patch_base_class):
+    stream = Users(config=MagicMock())
+
+    user_ids = set()
+    for record in stream.read_records(sync_mode=SyncMode.full_refresh):
```
Is there a need to fill the **_user_ids_** set before asserting that the set is of length 220? wouldn't the following snippet do the job? Might also save you some memory from allocating an additional data structure. 
```
records = stream.read_records(sync_mode=SyncMode.full_refresh):
records_length = sum(1 for record in records)
assert records_length == 220
```

lgomezm replied:
Updated in [fda250e](https://github.com/airbytehq/airbyte/pull/11452/commits/fda250ecf078624eff22eb745c3eaa7337c947ce).

lgomezm commented:
Hi @itaseskii. I've addressed your comments. Please take a look again when you get a chance and let me know if there's anything else I should do. Thanks for reviewing!

lgomezm commented:
This is the unit test execution:
![Screen Shot 2022-03-28 at 9 26 52 AM](https://user-images.githubusercontent.com/781929/160421498-fed8a46e-beda-4466-b97d-b7d8c88ca852.png)

itaseskii commented:
Hey @lgomezm I have went through you changes and I have no further comments, great job! :)

lgomezm commented:
@marcosmarxm @alafanechere Hi guys. This PR has been open for a while now. Could you take a look? Thanks in advance!

alafanechere commented on airbyte-integrations/connectors/source-notion/unit_tests/test_streams.py:
```diff
@@ -76,3 +78,56 @@ def test_backoff_time(patch_base_class):
     assert stream.backoff_time(response_mock) == expected_backoff_time
+
+
+def test_users_request_params(patch_base_class):
+    stream = Users(config=MagicMock())
+
+    # No next_page_token. First pull
+    inputs = {"stream_slice": None, "stream_state": None, "next_page_token": None}
+    expected_params = {"page_size": 100}
+    assert stream.request_params(**inputs) == expected_params
+
+    # When getting pages after the first pull.
+    inputs = {"stream_slice": None, "stream_state": None, "next_page_token": {"next_cursor": "123"}}
+    expected_params = {"start_cursor": "123", "page_size": 100}
+    assert stream.request_params(**inputs) == expected_params
```
nit: I suggest leverage parametrize wherever you can.
```suggestion
@pytest.mark.parametrize("next_page_token, expected_params", [(None, {"page_size": 100}), ({"next_cursor": "123"}, {"start_cursor": "123", "page_size": 100})])
def test_users_request_params(patch_base_class, next_page_token, expected_params):
    stream = Users(config=MagicMock())
    inputs = {"stream_slice": None, "stream_state": None, "next_page_token": next_page_token}
    assert stream.request_params(**inputs) == expected_params
```

alafanechere commented on airbyte-integrations/connectors/source-notion/unit_tests/test_streams.py:
```diff
@@ -76,3 +78,56 @@ def test_backoff_time(patch_base_class):
+    """
+    Test shows that Users stream uses pagination as per Notion API docs.
+    """
+
```
nit: I would suggest to use a function to generate your requests/responses sequence

alafanechere reviewed (approved):
Hi @lgomezm,
Thank you for your contribution. I made minor suggestions around the tests. I'll bump the version myself as I can commit to your branch.

alafanechere commented:
/test connector=connectors/source-notion

alafanechere commented:
/publish connector=connectors/source-notion

alafanechere commented:
/publish connector=connectors/source-notion auto-bump-version=false

alafanechere commented:
/publish connector=connectors/source-notion auto-bump-version=false

alafanechere reviewed (approved):
Thank you @lgomezm for this improvements. I had to update some schemas too as it looks like notion API evolved a bit.
